src/rdkafka_cgrp.c

```diff
@@ -2799,18 +3312,130 @@ static void rd_kafka_cgrp_check_unassign_done (rd_kafka_cgrp_t *rkcg,
 
 
 
+/**
+ * Call when all incremental unassign operations are done to transition to
+ * the next state.
+ */
+static void rd_kafka_cgrp_incr_unassign_done (rd_kafka_cgrp_t *rkcg,
+                                              const char *reason) {
+fprintf(stderr, ",,,, INCREMENTAL_UNASSIGN (done)\n");
+	rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "UNASSIGN",
+                "Group \"%s\" incremental unassign done in state %s "
+                "(join state %s): %s",
+                rkcg->rkcg_group_id->str,
+                rd_kafka_cgrp_state_names[rkcg->rkcg_state],
+                rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state],
+                reason);
+
+        /* if this action was underway when a terminate was initiated, it will
+         * be left to complete. now that's done, unassign all partitions */
+        if (rkcg->rkcg_flags & RD_KAFKA_CGRP_F_TERMINATE) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "UNASSIGN",
+                             "Group \"%s\" group is terminating, initiating "
+                             "full unassign",
+                             rkcg->rkcg_group_id->str);
+                rd_kafka_cgrp_unassign(rkcg);
+                return;
+        }
+
+        if (rkcg->rkcg_rebalance_incr_assignment) {
+
+                /** This incremental unassign was part of a normal rebalance
+                 *  (in which the revoke set was not empty). Immediately
+                 *  trigger the assign that follows this revoke. The protocol
+                 *  dictates this should occur even if the new assignment
+                 *  set is empty.
+                 *
+                 *  Also, since this rebalance had some revoked partitions,
+                 *  a re-join should occur following the assign.
+                 */
+                rkcg->rkcg_rebalance_rejoin = rd_true;
+                rd_kafka_rebalance_op_incr(
+                        rkcg,
+                        RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS,
+                        rkcg->rkcg_rebalance_incr_assignment,
+                        rd_true/*rejoin following assign*/,
+                        "cooperative assign after revoke");
+
+                rd_kafka_topic_partition_list_destroy(
+                        rkcg->rkcg_rebalance_incr_assignment);
+                rkcg->rkcg_rebalance_incr_assignment = NULL;
+
+                rd_kafka_cgrp_serve(rkcg);
```

Comment:
I put it here to ensure the rebalance op added immediately above it is actioned immediately.
it may not be required, i may have analyzed this incorrectly. 
anyway, i added a comment for now.

Reply:
Got it. I was just trying to make it fast, I didn't appreciate the recursive call.